File: app/src/main/java/com/omarea/scripts/switchs/SwitchConfigReader.java
package com.omarea.scripts.switchs;

import android.content.Context;
import android.util.Log;
import android.util.Xml;
import android.widget.Toast;

import com.omarea.scripts.ExtractAssets;
import com.omarea.scripts.simple.shell.ExecuteCommandWithOutput;

import org.xmlpull.v1.XmlPullParser;

import java.io.InputStream;
import java.util.ArrayList;
import java.util.Objects;

/**
 * Created by Hello on 2018/04/01.
 */

public class SwitchConfigReader {
    private static final String ASSETS_FILE = "file:///android_asset/";

    private static InputStream getConfig(Context context){
        try {
            return context.getAssets().open("switchs.xml");
        } catch (Exception ex) {
            return null;
        }
    }

    public static ArrayList<SwitchInfo> readActionConfigXml(Context context) {
        try {
            InputStream fileInputStream = getConfig(context);
            if (fileInputStream == null)
                return new ArrayList<>();
            XmlPullParser parser = Xml.newPullParser();// 获取xml解析器
            parser.setInput(fileInputStream, "utf-8");// 参数分别为输入流和字符编码
            int type = parser.getEventType();
            ArrayList<SwitchInfo> actions = null;
            SwitchInfo action = null;
            while (type != XmlPullParser.END_DOCUMENT) {// 如果事件不等于文档结束事件就继续循环
                switch (type) {
                    case XmlPullParser.START_TAG:
                        if ("switchs".equals(parser.getName())) {
                            actions = new ArrayList<>();
                        } else if ("switch".equals(parser.getName())) {
                            action = new SwitchInfo();
                            for (int i = 0; i < parser.getAttributeCount(); i++) {
                                switch (parser.getAttributeName(i)) {
                                    case "root": {
                                        action.root = Objects.equals(parser.getAttributeValue(i), "true");
                                        break;
                                    }
                                    case "confirm": {
                                        action.confirm = Objects.equals(parser.getAttributeValue(i), "true");
                                        break;
                                    }
                                    case "start": {
                                        action.start = parser.getAttributeValue(i);
                                        break;
                                    }
                                }
                            }
                        } else if (action != null) {
                            if ("title".equals(parser.getName())) {
                                action.title = parser.nextText();
                            } else if ("desc".equals(parser.getName())) {
                                for (int i = 0; i < parser.getAttributeCount(); i++) {
                                    String attrValue = parser.getAttributeValue(i);
                                    switch (parser.getAttributeName(i)) {
                                        case "su": {
                                            if (attrValue.trim().startsWith(ASSETS_FILE)) {
                                                String path = new ExtractAssets(context).extractToFilesDir(attrValue.trim());
                                                action.descPollingSUShell = "chmod 7777 " + path + "\n" + path;
                                            } else {
                                                action.descPollingSUShell = attrValue;
                                            }
                                            action.desc = executeResultRoot(context, action.descPollingSUShell);
                                            break;
                                        }
                                        case "sh": {
                                            if (attrValue.trim().startsWith(ASSETS_FILE)) {
                                                String path = new ExtractAssets(context).extractToFilesDir(attrValue.trim());
                                                action.descPollingShell = "chmod 7777 " + path + "\n" + path;
                                            } else {
                                                action.descPollingShell = attrValue;
                                            }
                                            action.desc = executeResultRoot(context, action.descPollingShell);
                                            break;
                                        }
                                    }
                                }
                                if (action.desc == null || action.desc.isEmpty())
                                    action.desc = parser.nextText();
                            } else if ("getstate".equals(parser.getName())) {
                                String script = parser.nextText();
                                if (script.trim().startsWith(ASSETS_FILE)) {
                                    action.getStateType = SwitchInfo.ActionScript.ASSETS_FILE;
                                    String path = new ExtractAssets(context).extractToFilesDir(script.trim());
                                    action.getState = "chmod 7777 " + path + "\n" + path;
                                } else {
                                    action.getState = script;
                                }
                            } else if ("setstate".equals(parser.getName())) {
                                String script = parser.nextText();
                                if (script.trim().startsWith(ASSETS_FILE)) {
                                    action.setStateType = SwitchInfo.ActionScript.ASSETS_FILE;
                                    String path = new ExtractAssets(context).extractToFilesDir(script.trim());
                                    action.setState = "chmod 7777 " + path + "\n" + path;
                                } else {
                                    action.setState = script;
                                }
                            }
                        }
                        break;
                    case XmlPullParser.END_TAG:
                        if ("switch".equals(parser.getName()) && actions != null && action != null) {
                            if (action.title == null) {
                                action.title = "";
                            }
                            if (action.desc == null) {
                                action.desc = "";
                            }
                            if (action.getState == null) {
                                action.getState = "";
                            } else {
                                String shellResult = ExecuteCommandWithOutput.executeCommandWithOutput(action.root, action.getState);
                                action.selected = shellResult != null && (shellResult.equals("1") || shellResult.toLowerCase().equals("true"));
                            }
                            if (action.setState == null) {
                                action.setState = "";
                            }

                            actions.add(action);
                            action = null;
                        }
                        break;
                }
                type = parser.next();// 继续下一个事件
            }

            return actions;
        } catch (Exception ex) {
            Toast.makeText(context, ex.getMessage(), Toast.LENGTH_LONG).show();
            Log.d("VTools ReadConfig Fail！", ex.getMessage());
        }
        return null;
    }

    private static String executeResult(Context context, String script) {
        if (script.trim().startsWith(ASSETS_FILE)) {
            String path = new ExtractAssets(context).extractToFilesDir(script.trim());
            script = "chmod 7777 " + path + "\n" + path;
        }
        String shellResult = ExecuteCommandWithOutput.executeCommandWithOutput(false, script);
        return shellResult;
    }

    private static String executeResultRoot(Context context, String script) {
        if (script.trim().startsWith(ASSETS_FILE)) {
            String path = new ExtractAssets(context).extractToFilesDir(script.trim());
            script = "chmod 7777 " + path + "\n" + path;
        }
        String shellResult = ExecuteCommandWithOutput.executeCommandWithOutput(true, script);
        return shellResult;
    }
}

File: app/src/main/java/com/omarea/scripts/switchs/SwitchConfigReader.kt
package com.omarea.scripts.switchs

import android.content.Context
import android.util.Log
import android.util.Xml
import android.widget.Toast
import com.omarea.scripts.ExtractAssets
import com.omarea.shell.KeepShellPublic
import org.xmlpull.v1.XmlPullParser
import java.io.InputStream
import java.util.*

/**
 * Created by Hello on 2018/04/01.
 */

object SwitchConfigReader {
    private const val ASSETS_FILE = "file:///android_asset/"

    private fun getConfig(context: Context): InputStream? {
        try {
            return context.assets.open("switchs.xml")
        } catch (ex: Exception) {
            return null
        }

    }

    fun readActionConfigXml(context: Context): ArrayList<SwitchInfo>? {
        try {
            val fileInputStream = getConfig(context) ?: return ArrayList()
            val parser = Xml.newPullParser()// 获取xml解析器
            parser.setInput(fileInputStream, "utf-8")// 参数分别为输入流和字符编码
            var type = parser.eventType
            var actions: ArrayList<SwitchInfo>? = null
            var action: SwitchInfo? = null
            while (type != XmlPullParser.END_DOCUMENT) {// 如果事件不等于文档结束事件就继续循环
                when (type) {
                    XmlPullParser.START_TAG -> if ("switchs" == parser.name) {
                        actions = ArrayList()
                    } else if ("switch" == parser.name) {
                        action = SwitchInfo()
                        for (i in 0 until parser.attributeCount) {
                            when (parser.getAttributeName(i)) {
                                "root" -> {
                                    action.root = parser.getAttributeValue(i) == "true"
                                }
                                "confirm" -> {
                                    action.confirm = parser.getAttributeValue(i) == "true"
                                }
                                "start" -> {
                                    action.start = parser.getAttributeValue(i)
                                }
                            }
                        }
                    } else if (action != null) {
                        if ("title" == parser.name) {
                            action.title = parser.nextText()
                        } else if ("desc" == parser.name) {
                            for (i in 0 until parser.attributeCount) {
                                val attrValue = parser.getAttributeValue(i)
                                when (parser.getAttributeName(i)) {
                                    "su" -> {
                                        if (attrValue.trim { it <= ' ' }.startsWith(ASSETS_FILE)) {
                                            val path = ExtractAssets(context).extractToFilesDir(attrValue.trim { it <= ' ' })
                                            action.descPollingSUShell = "chmod 0755 $path\n$path"
                                        } else {
                                            action.descPollingSUShell = attrValue
                                        }
                                        action.desc = executeResultRoot(context, action.descPollingSUShell)
                                    }
                                    "sh" -> {
                                        if (attrValue.trim { it <= ' ' }.startsWith(ASSETS_FILE)) {
                                            val path = ExtractAssets(context).extractToFilesDir(attrValue.trim { it <= ' ' })
                                            action.descPollingShell = "chmod 0755 $path\n$path"
                                        } else {
                                            action.descPollingShell = attrValue
                                        }
                                        action.desc = executeResultRoot(context, action.descPollingShell)
                                    }
                                }
                            }
                            if (action.desc == null || action.desc.isEmpty())
                                action.desc = parser.nextText()
                        } else if ("getstate" == parser.name) {
                            val script = parser.nextText()
                            if (script.trim { it <= ' ' }.startsWith(ASSETS_FILE)) {
                                action.getStateType = SwitchInfo.ActionScript.ASSETS_FILE
                                val path = ExtractAssets(context).extractToFilesDir(script.trim { it <= ' ' })
                                action.getState = "chmod 0755 $path\n$path"
                            } else {
                                action.getState = script
                            }
                        } else if ("setstate" == parser.name) {
                            val script = parser.nextText()
                            if (script.trim { it <= ' ' }.startsWith(ASSETS_FILE)) {
                                action.setStateType = SwitchInfo.ActionScript.ASSETS_FILE
                                val path = ExtractAssets(context).extractToFilesDir(script.trim { it <= ' ' })
                                action.setState = "chmod 0755 $path\n$path"
                            } else {
                                action.setState = script
                            }
                        }
                    }
                    XmlPullParser.END_TAG -> if ("switch" == parser.name && actions != null && action != null) {
                        if (action.title == null) {
                            action.title = ""
                        }
                        if (action.desc == null) {
                            action.desc = ""
                        }
                        if (action.getState == null) {
                            action.getState = ""
                        } else {
                            val shellResult = KeepShellPublic.doCmdSync(action.getState)
                            action.selected = shellResult != "error" && (shellResult == "1" || shellResult.toLowerCase() == "true")
                        }
                        if (action.setState == null) {
                            action.setState = ""
                        }

                        actions.add(action)
                        action = null
                    }
                }
                type = parser.next()// 继续下一个事件
            }

            return actions
        } catch (ex: Exception) {
            Toast.makeText(context, ex.message, Toast.LENGTH_LONG).show()
            Log.d("VTools ReadConfig Fail！", ex.message)
        }

        return null
    }

    private fun executeResult(context: Context, script: String): String {
        var script = script
        if (script.trim { it <= ' ' }.startsWith(ASSETS_FILE)) {
            val path = ExtractAssets(context).extractToFilesDir(script.trim { it <= ' ' })
            script = "chmod 0755 $path\n$path"
        }
        return KeepShellPublic.doCmdSync(script)
    }

    private fun executeResultRoot(context: Context, script: String): String {
        var script = script
        if (script.trim { it <= ' ' }.startsWith(ASSETS_FILE)) {
            val path = ExtractAssets(context).extractToFilesDir(script.trim { it <= ' ' })
            script = "chmod 0755 $path\n$path"
        }
        return KeepShellPublic.doCmdSync(script)
    }
}
